src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head = new Node<>(null, value, null);
+            tail = head;
+        } else if (tail.next == null) {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        } else {
+            tail.next = new Node<>(tail, value, null);
+            tail = tail.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Index " + index + " is out of size: " + size);
+        }
+        if (index == 0 && head == null || index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+            size++;
+            return;
+        }
+        Node<T> indexNode = getNodeByIndex(index - 1);
+        indexNode.next = new Node<>(indexNode, value, indexNode.next);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        if (isEmpty()) {
+            return null;
+        }
+        Node<T> result = getNodeByIndex(index);
+        return result.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        if (isEmpty()) {
+            return null;
+        }
+        Node<T> indexNode = getNodeByIndex(index);
+        T result = indexNode.value;
+        indexNode.value = value;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (isEmpty()) {
+            return null;
+        }
+        if (index == 0) {
+            return removeFirst();
+        }
+        Node<T> indexNode = getNodeByIndex(index);
+        unlink(indexNode);
+        return indexNode.value;
```

Comment:
unlink contains remove first. Also `checkIndex()` will be inside `getNodeByIndex()` that's why you dont need to call it 
```suggestion
        Node<T> indexNode = getNodeByIndex(index);
        unlink(indexNode);
        return indexNode.value;
```